cli/src/commands/metadata/retain.rs

```diff
@@ -0,0 +1,603 @@
+                vf.output.id
+            });
+
+            pallet_storage_types
+                .chain(pallet_constant_types)
+                .chain(pallet_event_type)
+                .chain(pallet_error_type)
+                .chain(pallet_view_fn_outputs)
+        });
+
+        // Runtime API outputs (inputs don't matter for stripping):
+        let runtime_api_outputs = self.apis.iter().flat_map(|api| {
+            &api.methods
+        }).map(|method| {
+            method.output.id
+        });
+
+        // Transaction Extension types:
+        let transaction_extension_types = self.extrinsic.transaction_extensions.iter().flat_map(|ext| {
+            [ext.ty.id, ext.implicit.id].into_iter()
+        });
+
+        // Custom types:
+        let custom_type_ids = self.custom.map.values().map(|value| {
+            value.ty.id
+        });
+
+        // Subxt needs this type so we always ensure to preserve it 
+        // even if it's not explicitly mentioned anywhere:
+        let dispatch_error_ty = std::iter::once_with(|| {
+            self
+                .types
+                .types
+                .iter()
+                .find(|ty| ty.ty.path.segments == ["sp_runtime", "DispatchError"])
+                .expect("Metadata must contain sp_runtime::DispatchError")
+                .id
+        });
+
+        // Return all IDs gathered:
+        pallet_types
+            .chain(runtime_api_outputs)
+            .chain(transaction_extension_types)
+            .chain(custom_type_ids)
+            .chain(dispatch_error_ty)
+    }
+
+    fn iterate_type_ids_ignored_for_stripping(&self) -> impl Iterator<Item=u32> {
+        // Pallet calls:
+        let pallet_types = self.pallets.iter().flat_map(|pallet| {
+            // We don't need to _decode_ any outer enums used in calls.
+            let pallet_call_types = pallet.calls.as_ref().into_iter().map(|calls| {
+                calls.ty.id
+            });
+
+            // Likewise with pallet view function input types.
+            let pallet_view_fn_inputs = pallet.view_functions.iter().flat_map(|vf| {
+                vf.inputs.iter().map(|input| input.ty.id)
+            });
+
+            pallet_call_types.chain(pallet_view_fn_inputs)
+        });
+
+        // Runtime API inputs:
+        let runtime_api_inputs = self.apis.iter().flat_map(|api| {
+            &api.methods
+        }).flat_map(|method| {
+            method.inputs.iter().map(|input| input.ty.id)
+        });
+
+        // The extrinsic type IDs:
+        let extrinsic_type_ids = [
+            self.extrinsic.address_ty.id,
+            self.extrinsic.signature_ty.id
+        ];
+
+        pallet_types.chain(runtime_api_inputs).chain(extrinsic_type_ids)
+    }
+}
+
+/// This is implemented for any metadata version that a [`PortableRegistry`] can be obtained from.
+trait GetPortableRegistry {
+    fn get_portable_registry(&self) -> &PortableRegistry;
```

Comment:
I think get and get_mut is sufficient here, the trait is called GetPortableRegistry ^